File: frontend/lib/formula.ts
// A small expression language for the error-propagation tool: parses a
// formula typed by hand (`2*pi*sqrt(L/g)`) and evaluates it together with its
// partial derivatives with respect to every variable in it.
//
// The derivatives are what error propagation needs, and they are computed by
// forward-mode automatic differentiation -- every node returns its value plus
// the gradient accumulated so far -- rather than by a finite difference. A
// numerical difference would need a step size per variable, and picking one
// badly is silently wrong (too large biases the slope, too small is eaten by
// floating-point cancellation). The chain rule applied node by node has no
// such knob and is exact to machine precision.

export class FormulaError extends Error {
  // Character offset the problem was found at, for pointing the user at it.
  readonly position: number;

  constructor(message: string, position: number) {
    super(message);
    this.name = "FormulaError";
    this.position = position;
  }
}

export type Node =
  | { type: "number"; value: number }
  | { type: "variable"; name: string }
  | { type: "constant"; name: string; value: number }
  | { type: "negate"; operand: Node }
  | { type: "binary"; op: "+" | "-" | "*" | "/" | "^"; left: Node; right: Node }
  | { type: "call"; name: string; args: Node[] };

export type ParsedFormula = {
  ast: Node;
  // Variable names in order of first appearance, so the input rows the UI
  // builds from this follow the formula as it reads.
  variables: string[];
};

// Values the parser resolves itself. A name in here is never a variable, so
// `pi` and `e` cannot be given a value or an uncertainty of their own.
const CONSTANTS: Record<string, number> = {
  pi: Math.PI,
  e: Math.E,
};

// Each function carries its own derivative. `derivatives[i]` is ∂f/∂(arg i)
// evaluated at the same arguments, which is all the chain rule needs.
type FunctionSpec = {
  arity: number;
  evaluate: (args: number[]) => number;
  derivatives: ((args: number[]) => number)[];
};

const FUNCTIONS: Record<string, FunctionSpec> = {
  sqrt: {
    arity: 1,
    evaluate: ([x]) => Math.sqrt(x),
    derivatives: [([x]) => 1 / (2 * Math.sqrt(x))],
  },
  exp: {
    arity: 1,
    evaluate: ([x]) => Math.exp(x),
    derivatives: [([x]) => Math.exp(x)],
  },
  ln: {
    arity: 1,
    evaluate: ([x]) => Math.log(x),
    derivatives: [([x]) => 1 / x],
  },
  log10: {
    arity: 1,
    evaluate: ([x]) => Math.log10(x),
    derivatives: [([x]) => 1 / (x * Math.LN10)],
  },
  sin: {
    arity: 1,
    evaluate: ([x]) => Math.sin(x),
    derivatives: [([x]) => Math.cos(x)],
  },
  cos: {
    arity: 1,
    evaluate: ([x]) => Math.cos(x),
    derivatives: [([x]) => -Math.sin(x)],
  },
  tan: {
    arity: 1,
    evaluate: ([x]) => Math.tan(x),
    derivatives: [([x]) => 1 / Math.cos(x) ** 2],
  },
  asin: {
    arity: 1,
    evaluate: ([x]) => Math.asin(x),
    derivatives: [([x]) => 1 / Math.sqrt(1 - x ** 2)],
  },
  acos: {
    arity: 1,
    evaluate: ([x]) => Math.acos(x),
    derivatives: [([x]) => -1 / Math.sqrt(1 - x ** 2)],
  },
  atan: {
    arity: 1,
    evaluate: ([x]) => Math.atan(x),
    derivatives: [([x]) => 1 / (1 + x ** 2)],
  },
  atan2: {
    arity: 2,
    evaluate: ([y, x]) => Math.atan2(y, x),
    derivatives: [
      ([y, x]) => x / (x ** 2 + y ** 2),
      ([y, x]) => -y / (x ** 2 + y ** 2),
    ],
  },
  sinh: {
    arity: 1,
    evaluate: ([x]) => Math.sinh(x),
    derivatives: [([x]) => Math.cosh(x)],
  },
  cosh: {
    arity: 1,
    evaluate: ([x]) => Math.cosh(x),
    derivatives: [([x]) => Math.sinh(x)],
  },
  tanh: {
    arity: 1,
    evaluate: ([x]) => Math.tanh(x),
    derivatives: [([x]) => 1 / Math.cosh(x) ** 2],
  },
  abs: {
    arity: 1,
    evaluate: ([x]) => Math.abs(x),
    derivatives: [([x]) => Math.sign(x)],
  },
};

// `log` is deliberately not a function: it means log10 in most lab handbooks
// and the natural log in most programming languages, and silently picking a
// base is worse than asking which one was meant.
const AMBIGUOUS_LOG =
  "log は底が曖昧です。ln（自然対数）か log10 と書いてください";

export const functionNames = Object.keys(FUNCTIONS);
export const constantNames = Object.keys(CONSTANTS);

type Token = {
  kind: "number" | "name" | "operator" | "paren" | "comma";
  text: string;
  position: number;
  value?: number;
};

const OPERATORS = new Set(["+", "-", "*", "/", "^"]);

function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  while (i < source.length) {
    const char = source[i];

    if (/\s/.test(char)) {
      i += 1;
      continue;
    }

    if (/[0-9.]/.test(char)) {
      const start = i;
      while (i < source.length && /[0-9.]/.test(source[i])) i += 1;
      // Exponent notation, but only when digits actually follow: `1.6e-19`
      // is one number, while the `e` in `2*e` is Euler's constant and the
      // one in `2e` is a typo the parser should complain about rather than
      // swallow.
      const exponentSign = /[+-]/.test(source[i + 1] ?? "") ? 1 : 0;
      if (
        /[eE]/.test(source[i] ?? "") &&
        /[0-9]/.test(source[i + 1 + exponentSign] ?? "")
      ) {
        i += 1 + exponentSign;
        while (i < source.length && /[0-9]/.test(source[i])) i += 1;
      }
      const text = source.slice(start, i);
      const value = Number(text);
      if (!Number.isFinite(value)) {
        throw new FormulaError(`数値として読めません: ${text}`, start);
      }
      tokens.push({ kind: "number", text, position: start, value });
      continue;
    }

    // Unicode letters, so Greek names physicists actually write (θ, λ, Δx)
    // are variables like any other.
    if (/[\p{L}_]/u.test(char)) {
      const start = i;
      while (i < source.length && /[\p{L}\p{N}_]/u.test(source[i])) i += 1;
      tokens.push({
        kind: "name",
        text: source.slice(start, i),
        position: start,
      });
      continue;
    }

    if (OPERATORS.has(char)) {
      tokens.push({ kind: "operator", text: char, position: i });
      i += 1;
      continue;
    }

    if (char === "(" || char === ")") {
      tokens.push({ kind: "paren", text: char, position: i });
      i += 1;
      continue;
    }

    if (char === ",") {
      tokens.push({ kind: "comma", text: char, position: i });
      i += 1;
      continue;
    }

    throw new FormulaError(`使えない文字です: ${char}`, i);
  }

  return tokens;
}

// Recursive descent over the usual arithmetic precedence:
//
//   expression := term (('+' | '-') term)*
//   term       := unary (('*' | '/') unary)*
//   unary      := ('-' | '+') unary | power
//   power      := atom ('^' unary)?
//   atom       := number | name | name '(' args ')' | '(' expression ')'
//
// `^` is right-associative (a^b^c = a^(b^c)) and binds tighter than unary
// minus (-x^2 = -(x^2)), matching how the same formula reads on paper. Its
// right side is a unary so `x^-2` works.
class Parser {
  private index = 0;

  constructor(
    private readonly tokens: Token[],
    private readonly source: string,
    readonly variables: string[] = [],
  ) {}

  parse(): Node {
    if (this.tokens.length === 0) {
      throw new FormulaError("式が空です", 0);
    }
    const node = this.parseExpression();
    const leftover = this.peek();
    if (leftover) {
      throw new FormulaError(
        `式の続きが読めません: ${leftover.text}`,
        leftover.position,
      );
    }
    return node;
  }

  private peek(): Token | undefined {
    return this.tokens[this.index];
  }

  private next(): Token | undefined {
    return this.tokens[this.index++];
  }

  private endPosition(): number {
    return this.source.length;
  }

  private parseExpression(): Node {
    let left = this.parseTerm();
    for (;;) {
      const token = this.peek();
      if (
        token?.kind === "operator" &&
        (token.text === "+" || token.text === "-")
      ) {
        this.index += 1;
        left = {
          type: "binary",
          op: token.text,
          left,
          right: this.parseTerm(),
        };
        continue;
      }
      return left;
    }
  }

  private parseTerm(): Node {
    let left = this.parseUnary();
    for (;;) {
      const token = this.peek();
      if (
        token?.kind === "operator" &&
        (token.text === "*" || token.text === "/")
      ) {
        this.index += 1;
        left = {
          type: "binary",
          op: token.text,
          left,
          right: this.parseUnary(),
        };
        continue;
      }
      return left;
    }
  }

  private parseUnary(): Node {
    const token = this.peek();
    if (
      token?.kind === "operator" &&
      (token.text === "-" || token.text === "+")
    ) {
      this.index += 1;
      const operand = this.parseUnary();
      return token.text === "-" ? { type: "negate", operand } : operand;
    }
    return this.parsePower();
  }

  private parsePower(): Node {
    const base = this.parseAtom();
    const token = this.peek();
    if (token?.kind === "operator" && token.text === "^") {
      this.index += 1;
      return { type: "binary", op: "^", left: base, right: this.parseUnary() };
    }
    return base;
  }

  private parseAtom(): Node {
    const token = this.next();
    if (!token) {
      throw new FormulaError("式が途中で終わっています", this.endPosition());
    }

    if (token.kind === "number") {
      return { type: "number", value: token.value as number };
    }

    if (token.kind === "paren" && token.text === "(") {
      const inner = this.parseExpression();
      const closing = this.next();
      if (!closing || closing.kind !== "paren" || closing.text !== ")") {
        throw new FormulaError(
          "閉じ括弧が足りません",
          closing?.position ?? this.endPosition(),
        );
      }
      return inner;
    }

    if (token.kind === "name") {
      const following = this.peek();
      const isCall = following?.kind === "paren" && following.text === "(";

      if (isCall) {
        return this.parseCall(token);
      }

      if (Object.hasOwn(CONSTANTS, token.text)) {
        return {
          type: "constant",
          name: token.text,
          value: CONSTANTS[token.text],
        };
      }
      if (Object.hasOwn(FUNCTIONS, token.text)) {
        throw new FormulaError(
          `${token.text} は関数です。${token.text}(...) のように括弧が要ります`,
          token.position,
        );
      }
      if (token.text === "log") {
        throw new FormulaError(AMBIGUOUS_LOG, token.position);
      }
      if (!this.variables.includes(token.text)) {
        this.variables.push(token.text);
      }
      return { type: "variable", name: token.text };
    }

    throw new FormulaError(`ここには置けません: ${token.text}`, token.position);
  }

  private parseCall(name: Token): Node {
    const spec = Object.hasOwn(FUNCTIONS, name.text)
      ? FUNCTIONS[name.text]
      : undefined;
    if (!spec && name.text === "log") {
      throw new FormulaError(AMBIGUOUS_LOG, name.position);
    }
    if (!spec) {
      throw new FormulaError(
        `知らない関数です: ${name.text}（使えるのは ${functionNames.join(", ")}）`,
        name.position,
      );
    }

    this.index += 1; // consume "("
    const args: Node[] = [];
    if (this.peek()?.kind === "paren" && this.peek()?.text === ")") {
      this.index += 1;
    } else {
      for (;;) {
        args.push(this.parseExpression());
        const separator = this.next();
        if (separator?.kind === "comma") continue;
        if (separator?.kind === "paren" && separator.text === ")") break;
        throw new FormulaError(
          "引数の区切り（,）か閉じ括弧がありません",
          separator?.position ?? this.endPosition(),
        );
      }
    }

    if (args.length !== spec.arity) {
      throw new FormulaError(
        `${name.text} は引数が ${spec.arity} 個です（${args.length} 個ありました）`,
        name.position,
      );
    }
    return { type: "call", name: name.text, args };
  }
}

export function parseFormula(source: string): ParsedFormula {
  const parser = new Parser(tokenize(source), source);
  const ast = parser.parse();
  return { ast, variables: parser.variables };
}

export type Evaluation = {
  value: number;
  // ∂(formula)/∂name for every variable in the formula.
  gradient: Record<string, number>;
};

// Forward-mode automatic differentiation: each node reports its value and the
// partial derivative of that value with respect to every variable seen so
// far. Combining two nodes applies the ordinary product/quotient/chain rules
// to both halves at once.
export function evaluateWithGradient(
  ast: Node,
  values: Record<string, number>,
): Evaluation {
  return visit(ast, values);
}

function visit(node: Node, values: Record<string, number>): Evaluation {
  switch (node.type) {
    case "number":
      return { value: node.value, gradient: emptyGradient() };

    case "constant":
      return { value: node.value, gradient: emptyGradient() };

    case "variable": {
      const value = Object.hasOwn(values, node.name)
        ? values[node.name]
        : undefined;
      if (value === undefined) {
        throw new FormulaError(`${node.name} の値がありません`, 0);
      }
      // ∂x/∂x = 1; every other variable's partial stays 0 by omission.
      const gradient = emptyGradient();
      gradient[node.name] = 1;
      return { value, gradient };
    }

    case "negate": {
      const operand = visit(node.operand, values);
      return {
        value: -operand.value,
        gradient: scale(operand.gradient, -1),
      };
    }

    case "binary":
      return visitBinary(node, values);

    case "call": {
      const spec = FUNCTIONS[node.name];
      const args = node.args.map((arg) => visit(arg, values));
      const argValues = args.map((arg) => arg.value);
      let gradient = emptyGradient();
      args.forEach((arg, i) => {
        gradient = add(
          gradient,
          scale(arg.gradient, spec.derivatives[i](argValues)),
        );
      });
      return { value: spec.evaluate(argValues), gradient };
    }
  }
}

function visitBinary(
  node: Extract<Node, { type: "binary" }>,
  values: Record<string, number>,
): Evaluation {
  const left = visit(node.left, values);
  const right = visit(node.right, values);

  switch (node.op) {
    case "+":
      return {
        value: left.value + right.value,
        gradient: add(left.gradient, right.gradient),
      };
    case "-":
      return {
        value: left.value - right.value,
        gradient: add(left.gradient, scale(right.gradient, -1)),
      };
    case "*":
      return {
        value: left.value * right.value,
        gradient: add(
          scale(left.gradient, right.value),
          scale(right.gradient, left.value),
        ),
      };
    case "/":
      return {
        value: left.value / right.value,
        gradient: add(
          scale(left.gradient, 1 / right.value),
          scale(right.gradient, -left.value / right.value ** 2),
        ),
      };
    case "^": {
      const value = Math.pow(left.value, right.value);
      // d/dx x^n = n·x^(n-1) for the base, and d/dn x^n = x^n·ln x for the
      // exponent. The second term only exists when the exponent itself
      // depends on a variable (`x^n` with n measured); for a constant
      // exponent its gradient is empty and the ln disappears with it, which
      // matters because ln x is not defined for a negative base.
      let gradient = scale(
        left.gradient,
        right.value * Math.pow(left.value, right.value - 1),
      );
      if (Object.keys(right.gradient).length > 0) {
        gradient = add(
          gradient,
          scale(right.gradient, value * Math.log(left.value)),
        );
      }
      return { value, gradient };
    }
  }
}

// A gradient is keyed by variable names the user typed, so it is built
// without a prototype: `constructor` must not resolve to something inherited,
// and `__proto__ = 1` on a plain object silently sets the prototype instead
// of storing the partial, which would drop that variable from the result.
function emptyGradient(): Record<string, number> {
  return Object.create(null) as Record<string, number>;
}

function scale(
  gradient: Record<string, number>,
  factor: number,
): Record<string, number> {
  const scaled = emptyGradient();
  for (const [name, partial] of Object.entries(gradient)) {
    scaled[name] = partial * factor;
  }
  return scaled;
}

function add(
  a: Record<string, number>,
  b: Record<string, number>,
): Record<string, number> {
  const sum = emptyGradient();
  Object.assign(sum, a);
  for (const [name, partial] of Object.entries(b)) {
    sum[name] = (sum[name] ?? 0) + partial;
  }
  return sum;
}

File: backend/worker/app/formula.py
"""A small expression language for formulas typed by hand, e.g.
``A*exp(-x/tau)*cos(w*x + phi)``.

This is the Python half of the grammar the frontend's error-propagation tool
already speaks (``frontend/lib/formula.ts``). The two are kept in step by
``tests/fixtures/formula_grammar.json``, which both test suites read: a
formula that parses one way in the browser has to parse the same way here,
or the same text typed into two parts of the same app would mean two
different things.

**Python's own ``ast`` module cannot be used for this.** ``^`` is the power
operator here, as it is on paper and in the error-propagation tool, while
``ast.parse`` reads it as bitwise XOR -- so ``x^2`` would parse happily and
silently mean something else. Hence a parser written out by hand, which also
means no ``eval`` anywhere near user input.

Unlike the TypeScript version there is no automatic differentiation here:
the only caller is the curve fitter, and SciPy estimates its own Jacobian
numerically. What this does instead is evaluate over numpy arrays, so a
formula is applied to a whole column at once.
"""

import math
import re
from dataclasses import dataclass
from typing import Any

import numpy as np


class FormulaError(Exception):
    """A formula that cannot be parsed or evaluated.

    ``position`` is the character offset the problem was found at, so a
    caller can point at it; -1 when the problem is not tied to one spot.
    """

    def __init__(self, message: str, position: int = -1):
        super().__init__(message)
        self.position = position


# Resolved by the parser itself, so a name in here is never a variable and
# so can never become a fit parameter.
CONSTANTS: dict[str, float] = {
    "pi": math.pi,
    "e": math.e,
}

# numpy equivalents of the frontend's Math.* calls, chosen so they apply
# element-wise to an array. Arities match FUNCTIONS in formula.ts; the
# fixture test is what keeps them matching.
FUNCTIONS: dict[str, tuple[int, Any]] = {
    "sqrt": (1, np.sqrt),
    "exp": (1, np.exp),
    "ln": (1, np.log),
    "log10": (1, np.log10),
    "sin": (1, np.sin),
    "cos": (1, np.cos),
    "tan": (1, np.tan),
    "asin": (1, np.arcsin),
    "acos": (1, np.arccos),
    "atan": (1, np.arctan),
    "atan2": (2, np.arctan2),
    "sinh": (1, np.sinh),
    "cosh": (1, np.cosh),
    "tanh": (1, np.tanh),
    "abs": (1, np.abs),
}

OPERATORS = frozenset("+-*/^")

_NUMBER_START = re.compile(r"[0-9.]")
_DIGIT = re.compile(r"[0-9]")
# Unicode letters, so the Greek names physicists actually write (θ, λ, Δx)
# are variables like any other.
_NAME_START = re.compile(r"[^\W\d]|_", re.UNICODE)
_NAME_PART = re.compile(r"\w", re.UNICODE)


@dataclass(frozen=True)
class Token:
    kind: str  # number | name | operator | paren | comma
    text: str
    position: int
    value: float | None = None


def tokenize(source: str) -> list[Token]:
    """Split a formula into tokens, mirroring formula.ts's tokenizer."""
    tokens: list[Token] = []
    i = 0
    n = len(source)

    while i < n:
        char = source[i]

        if char.isspace():
            i += 1
            continue

        if _NUMBER_START.match(char):
            start = i
            while i < n and _NUMBER_START.match(source[i]):
                i += 1
            # Exponent notation, but only when digits actually follow:
            # `1.6e-19` is one number, while the `e` in `2*e` is Euler's
            # constant and the one in `2e` is a typo worth complaining
            # about rather than swallowing.
            following = source[i + 1] if i + 1 < n else ""
            exponent_sign = 1 if following in "+-" else 0
            after_sign = source[i + 1 + exponent_sign] if i + 1 + exponent_sign < n else ""
            if i < n and source[i] in "eE" and _DIGIT.match(after_sign or ""):
                i += 1 + exponent_sign
                while i < n and _DIGIT.match(source[i]):
                    i += 1
            text = source[start:i]
            try:
                value = float(text)
            except ValueError as exc:
                raise FormulaError(f"数値として読めません: {text}", start) from exc
            if not math.isfinite(value):
                raise FormulaError(f"数値として読めません: {text}", start)
            tokens.append(Token("number", text, start, value))
            continue

        if _NAME_START.match(char):
            start = i
            while i < n and _NAME_PART.match(source[i]):
                i += 1
            tokens.append(Token("name", source[start:i], start))
            continue

        if char in OPERATORS:
            tokens.append(Token("operator", char, i))
            i += 1
            continue

        if char in "()":
            tokens.append(Token("paren", char, i))
            i += 1
            continue

        if char == ",":
            tokens.append(Token("comma", char, i))
            i += 1
            continue

        raise FormulaError(f"使えない文字です: {char}", i)

    return tokens


# Node is a tagged tuple tree rather than a class hierarchy: it is only ever
# built here and walked by _evaluate, and the shape mirrors formula.ts's
# discriminated union one for one.
#
#   ("number", value)
#   ("constant", name, value)
#   ("variable", name)
#   ("negate", operand)
#   ("binary", op, left, right)
#   ("call", name, args)
Node = tuple


class _Parser:
    """Recursive descent over the usual arithmetic precedence:

        expression := term (('+' | '-') term)*
        term       := unary (('*' | '/') unary)*
        unary      := ('-' | '+') unary | power
        power      := atom ('^' unary)?
        atom       := number | name | name '(' args ')' | '(' expression ')'

    ``^`` is right-associative (a^b^c = a^(b^c)) and binds tighter than
    unary minus (-x^2 = -(x^2)), matching how the same formula reads on
    paper. Its right side is a unary so ``x^-2`` works.
    """

    def __init__(self, tokens: list[Token], source: str):
        self.tokens = tokens
        self.source = source
        self.index = 0
        # First-appearance order, so the inputs a caller builds from this
        # follow the formula as it reads.
        self.variables: list[str] = []

    def parse(self) -> Node:
        if not self.tokens:
            raise FormulaError("式が空です", 0)
        node = self.parse_expression()
        token = self.peek()
        if token is not None:
            raise FormulaError(f"余分な入力です: {token.text}", token.position)
        return node

    def peek(self) -> Token | None:
        return self.tokens[self.index] if self.index < len(self.tokens) else None

    def next(self) -> Token:
        token = self.peek()
        if token is None:
            raise FormulaError("式が途中で終わっています", len(self.source))
        self.index += 1
        return token

    def parse_expression(self) -> Node:
        node = self.parse_term()
        while True:
            token = self.peek()
            if token is not None and token.kind == "operator" and token.text in "+-":
                self.index += 1
                node = ("binary", token.text, node, self.parse_term())
            else:
                return node

    def parse_term(self) -> Node:
        node = self.parse_unary()
        while True:
            token = self.peek()
            if token is not None and token.kind == "operator" and token.text in "*/":
                self.index += 1
                node = ("binary", token.text, node, self.parse_unary())
            else:
                return node

    def parse_unary(self) -> Node:
        token = self.peek()
        if token is not None and token.kind == "operator" and token.text in "+-":
            self.index += 1
            operand = self.parse_unary()
            return ("negate", operand) if token.text == "-" else operand
        return self.parse_power()

    def parse_power(self) -> Node:
        base = self.parse_atom()
        token = self.peek()
        if token is not None and token.kind == "operator" and token.text == "^":
            self.index += 1
            return ("binary", "^", base, self.parse_unary())
        return base

    def parse_atom(self) -> Node:
        token = self.next()

        if token.kind == "number":
            assert token.value is not None
            return ("number", token.value)

        if token.kind == "name":
            following = self.peek()
            if following is not None and following.kind == "paren" and following.text == "(":
                return self.parse_call(token)
            if token.text in CONSTANTS:
                return ("constant", token.text, CONSTANTS[token.text])
            if token.text in FUNCTIONS:
                raise FormulaError(
                    f"{token.text} は関数です。{token.text}(...) と書いてください",
                    token.position,
                )
            if token.text not in self.variables:
                self.variables.append(token.text)
            return ("variable", token.text)

        if token.kind == "paren" and token.text == "(":
            node = self.parse_expression()
            closing = self.peek()
            if closing is None or closing.kind != "paren" or closing.text != ")":
                # Pointed at where the ')' should have been, not at the '('
                # that opened it -- which is where formula.ts points too.
                raise FormulaError(
                    "閉じ括弧がありません",
                    closing.position if closing else len(self.source),
                )
            self.index += 1
            return node

        raise FormulaError(f"ここには書けません: {token.text}", token.position)

    def parse_call(self, name_token: Token) -> Node:
        spec = FUNCTIONS.get(name_token.text)
        if spec is None:
            raise FormulaError(f"知らない関数です: {name_token.text}", name_token.position)
        arity, _ = spec

        self.index += 1  # consume '('
        args: list[Node] = []
        closing = self.peek()
        if closing is not None and closing.kind == "paren" and closing.text == ")":
            self.index += 1
        else:
            while True:
                args.append(self.parse_expression())
                separator = self.peek()
                if separator is not None and separator.kind == "comma":
                    self.index += 1
                    continue
                if separator is not None and separator.kind == "paren" and separator.text == ")":
                    self.index += 1
                    break
                raise FormulaError(
                    "引数の区切りか閉じ括弧が必要です",
                    separator.position if separator else len(self.source),
                )

        if len(args) != arity:
            raise FormulaError(
                f"{name_token.text} は引数{arity}個です（{len(args)}個でした）",
                name_token.position,
            )
        return ("call", name_token.text, args)


@dataclass(frozen=True)
class ParsedFormula:
    ast: Node
    # Variable names in first-appearance order.
    variables: tuple[str, ...]


def parse_formula(source: str) -> ParsedFormula:
    """Parse a formula, returning its tree and the variables it mentions.

    Names resolved as constants (``pi``, ``e``) or functions are not
    variables, so they can never be mistaken for something to fit.
    """
    parser = _Parser(tokenize(source), source)
    ast = parser.parse()
    return ParsedFormula(ast=ast, variables=tuple(parser.variables))


def evaluate(ast: Node, values: dict[str, Any]) -> Any:
    """Evaluate a parsed formula.

    Each value may be a scalar or a numpy array; arrays are applied
    element-wise, which is what lets a fitter evaluate the model over a
    whole x column at once. Mixing the two follows numpy's broadcasting.

    Non-finite results are not rejected here: a fitter walks through
    parameter values that may well be out of a function's domain on the way
    to a good answer, and numpy's nan is the right thing to hand back for
    that. Callers that need a finite answer check the one they end up with.
    """
    kind = ast[0]

    if kind in ("number", "constant"):
        return ast[-1]

    if kind == "variable":
        name = ast[1]
        if name not in values:
            raise FormulaError(f"{name} の値がありません")
        return values[name]

    if kind == "negate":
        return -evaluate(ast[1], values)

    if kind == "binary":
        _, op, left_node, right_node = ast
        left = evaluate(left_node, values)
        right = evaluate(right_node, values)
        # numpy's warnings for divide-by-zero, overflow and invalid powers
        # are suppressed rather than raised: they are ordinary events while
        # a fitter searches, and the nan/inf they produce is the answer.
        with np.errstate(all="ignore"):
            if op == "+":
                return left + right
            if op == "-":
                return left - right
            if op == "*":
                return left * right
            if op == "/":
                return np.divide(left, right)
            if op == "^":
                # float exponents of negative bases are nan, not an error,
                # matching JavaScript's ** and keeping the fitter moving.
                return np.power(np.asarray(left, dtype=float), right)
        raise FormulaError(f"知らない演算子です: {op}")

    if kind == "call":
        _, name, arg_nodes = ast
        _, fn = FUNCTIONS[name]
        args = [evaluate(node, values) for node in arg_nodes]
        with np.errstate(all="ignore"):
            return fn(*args)

    raise FormulaError(f"知らないノードです: {kind}")
